<style {font-family:Consolas,Monaco,Lucida Console,Liberation Mono,DejaVu Sans Mono,Bitstream Vera Sans Mono,Courier New, monospace;font-size:6}>

Intro
=====

The barcode data was analysed using mirdeep helper scripts and custom made scripts present here.

The analyses consisted of the following operations on the data:

1. Adapter trim
2. Read collapse
3. Linker trim
4. collapse again
5. Filter by length
6. Homolog removal 1mm
7. Homolog removal 2mm

improvements:
- merge files and do homolog removal


Summary table
=============
this table tries to sum it up

current:

step	|	stepname			|	sample		|	Reads	|	Collapsed reads
--------|---------------------------------------|-----------------------|---------------|----------------------
1	|	Adapter trim			|	17-eGFPfwd3E	|	843410/644126 (fwd/reverse)	|	
2	|	Collapse after adapter trim	|	17-eGFPfwd3E	|	1487536	|	93413
3	|	Trim on linkers			|	17-eGFPfwd3E	|	1395112	|	69513
4	|	Collapse after linker trim	|	17-eGFPfwd3E	|	1395112	|	65486
5	|	Filter on length		|	17-eGFPfwd3E	|	1297389	|	58390
6	|	Homolog removal 1mm		|	17-eGFPfwd3E	|	1183152	|	855 [pdf](https://github.com/mmterpstra/BA-BarcodeAnalysis/blob/master/pdf/17-eGFPfwd3E.collapse_md.keep_33-34.fa.pdf?raw=true)
1	|	Adapter trim			|	18-eGFPfwd3L	|	1481063/1119915 (fwd/reverse)	|	
2	|	Collapse after adapter trim	|	18-eGFPfwd3L	|	2600978	|	158669
3	|	Trim on linkers			|	18-eGFPfwd3L	|	2512249	|	134702
4	|	Collapse after linker trim	|	18-eGFPfwd3L	|	2512249	|	134210
5	|	Filter on length		|	18-eGFPfwd3L	|	2334138	|	119084
6	|	Homolog removal 1mm		|	18-eGFPfwd3L	|	2086164	|	1367 [pdf](https://github.com/mmterpstra/BA-BarcodeAnalysis/blob/master/pdf/18-eGFPfwd3L.collapse_md.keep_33-34.fa.pdf?raw=true)
1	|	Adapter trim			|	8-eGFPfwd3L	|	1543617/1562697 (fwd/reverse)	|	
2	|	Collapse after adapter trim	|	8-eGFPfwd3L	|	3106314	|	163385
3	|	Trim on linkers			|	8-eGFPfwd3L	|	3010420	|	139339
4	|	Collapse after linker trim	|	8-eGFPfwd3L	|	3010420	|	138903
5	|	Filter on length		|	8-eGFPfwd3L	|	2782738	|	118561
6	|	Homolog removal 1mm		|	8-eGFPfwd3L	|	2504569	|	1868[pdf](https://github.com/mmterpstra/BA-BarcodeAnalysis/blob/master/pdf/8-eGFPfwd3L.collapse_md.keep_33-34.fa.pdf?raw=true)
1	|	Adapter trim			|	eGFPfwd3-193	|	867406/1208614 (fwd/reverse)|
2	|	Collapse after adapter trim	|	eGFPfwd3-193	|	2076020	|	110927
3	|	Trim on linkers			|	eGFPfwd3-193	|	2013563	|	94335
4	|	Collapse after linker trim	|	eGFPfwd3-193	|	2013563	|	93985
5	|	Filter on length		|	eGFPfwd3-193	|	1862435	|	80481
6	|	Homolog removal 1mm		|	eGFPfwd3-193	|	1700850	|	1288 [pdf](https://github.com/mmterpstra/BA-BarcodeAnalysis/blob/master/pdf/eGFPfwd3-193.collapse_md.keep_33-34.fa.pdf?raw=true)



old:

1. sample after adapter trim | reads(fwd/reverse) | 2. after collapse reads | Reads | Collapsed reads | 3. remove linkers (GAC/TAAGG/TACCAGTAAGG) | Reads | Collapsed reads | % Reads vs 2. after collapse | 4. Filter on length >= 33 &  <= 34 | Reads | Collapsed reads | % Reads vs Reads at collapse | 5. Remove homologs | Reads | Collapsed reads
--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | ---
17-eGFPfwd3E | 843410/644126  | | 843410 | 60415 |  | 71968 | 1411 | 8.53% |  | 67202 | 1307 | 7.97% |  | 649390 | 640
18-eGFPfwd3L| 1481063/1119915 | | 1481063 | 106581 |  | 128694 | 1945 | 8.69% |  | 120046 | 1793 | 8.11% |  | 1177324 | 901
8-eGFPfwd3L| 1543617/1562697 | | 1543617 | 88881 |  | 137310 | 2407 | 8.90% |  | 127675 | 2219 | 8.27% |  | 1247783 | 742
eGFPfwd3-193| 867406/1208614 | | 867406 | 49302 |  | 77798 | 875 | 8.97% |  | 72331 | 793 | 8.34% |  | 717257 | 669

initital fastq: 13882978 reads


Adapter trim
============

A trim for the following adapters was performed. In our script allowed for 1 single nucleotide mismatch and it was done seperately for recognising the forward and reverse insetrion of the sequence. Note that in the actual contruct the reverse primer is reverse complemented to match the forward primer. 

PCRset | Forward primer name | Forward primer sequence | Forward primer sequence_original | Reverse primer name | Reverse primer sequence
--- | --- | --- | --- | --- | ---
1 | 8-eGFPfwd3L | AATCCGTCCAAGGCATGGACGAGCTGTACAAG | AATCCGTCCAAGGCATGGACGAGCTGTACAAG | BC-rev-L+2:      | ATGGGGGATCCTCACTGGCC
2 | 17-eGFPfwd3E | CAAGAATATTCTCGGCATGGACGAGCTGTACAAG | CAAGAATATTCTCGGCATGGACGAGCTG | BC-rev-L+6:   | TAATATGGGGGATCCTCACTGGCC
3 | eGFPfwd3-193 | TCATCTCTGGCATGGACGAGCTGTACAAG | TCATCTCTGGCATGGACGAGCTGTACAAG | BC-rev-L+5:   | AATATGGGGGATCCTCACTGGCC
4 | 18-eGFPfwd3L | ATCGAATTTATGGCATGGACGAGCTGTACAAG | ATCGAATTTATGGCATGGACGAGCTGTACAAG | BC-rev-L+1:   | TGGGGGATCCTCACTGGCC



Read collapse
=============

recution of duplicate reads to a single fasta entry.
fastq example:

```
@read1
AAAAAAAAAAAAAAAAAA
+read1
ABBDFDFDFDFDDDGGGG
@read2
AAAAAAAAAAAAAAAAAA
+read2
ABBDFDFDFDFDDDDDDD
@read3
AAAAAAAAAAAAAAANAA
+read4
ABBDFDFDFDFDDDDDDD
```

after collapse:
```
>SEQ_1_x2
AAAAAAAAAAAAAAAAAA
>SEQ_3_x1
AAAAAAAAAAAAAAANAA
```

Linker trim
===========

Removal of the linker sequences from the end of the reads (GAC/TAAGG/TACCAGTAAGG).
These linkers are present between the adapters and the 33nt barcode.


Filter by length
================

|keep 33-34: |
| --- |
|GTTACATAGTTGCCGAATTATCGCAGAATGTAA |
|AAGGTACCATGTCACCGGTTTAAACCATACTAAG |
|not |
|GTTACATAGT |
|GTTACATAGTAAGGTACCATGTCACCGGTTTAAACCATACTAAG |

Homolog removal
===============

Removal of similar sequences by trimming data.

this was done:
 - Allowing 1 single nucleotide mismatch
 - Removing the reads with a count less than or equal to 1
 - Removing reads when it has a homolog read with a higher count

Homolog removal (again)
===============

Removal of similar sequences by trimming data.

this was done:
 - Allowing **2** single nucleotide mismatches
 - Removing the reads with a count less than or equal to 1
 - Removing reads when it has a homolog read with a higher count

code for different tasks
========================

this is example code and will probably not work on your system.

Create plot

```sh
for t in $(ls *.tsv) ; do
	echo $t;
	head -1 $t >${t}.tmp;
	tail -n+2 $t|sort -d$(echo -e "\t") -k2,2nr  >>${t}.tmp;
	Rscript ../PcrProductKlaas/CreateHist.R ${t}.tmp;
	mv $(basename ${t} .tsv).tmp.pdf $(basename $t .tsv).pdf;
	rm ${t}.tmp ;
done
```

Create tsv

```sh
for fa in $(ls filterlength/*/bctrim.collapsed.fa); do 
	basenamefa=$(basename $(dirname $fa) .keep_33-34);dirnamefa=./;
	echo -e "ID\tcounts\tsequence" > ${dirnamefa}/${basenamefa}.tsv;
	perl -ne 'chomp; s/^>SEQ_|\n//g;s/_x|_/\t/g;
	print $_ if $.%2==1;print "\t".$_."\n" if $.%2==0;' $fa >> ${dirnamefa}/${basenamefa}.tsv;
done
```

Filter for homolog barcodes

```sh
(for i in $(ls filterlength/*.fa); do 
	echo $i;
	perl ~/workspace/FastqManipulations/barcodeCleanup.pl $i &
done)&
```

code for filtering by length

```sh
mkdir -p  filterlength; 
for fa in $(ls collapseTrimlinkers/*.fa); do 
	perl ~/workspace/FastqManipulations/FilterFaByLength.pl 33 34 $fa > filterlength/$(basename $fa .trim_GAC_TAAGG.fa).keep_33-34.fa ;
done
```

collapse again

```sh
mkdir -p collapseTrimlinkers
for fa in $(ls trimlinkers/*.fa); do 
	basenamefa=$(basename $fa .trim_GAC_TAAGG.fa);
	cat $fa | collapse_reads_md.pl - SEQ > collapseTrimlinkers/${basenamefa}.collapse_md.fa;
done
```
code for trimming linkers
```sh
mkdir -p trimlinkers;
for fa in $(ls collapse/*fa); do 
	basenamefa=$(basename $fa .collapse_md.fa);
	cat $fa | \
	perl ~/workspace/FastqManipulations/TrimGACfa.pl -| \
	perl ~/workspace/FastqManipulations/TrimTAAGGfa.pl - > trimlinkers/${basenamefa}.trim_GAC_TAAGG.fa;
done
```


code for collapse (using the mirdeep package)

```sh
mkdir -p collapse;
for fq in $(ls trimAdapters/*.fwtrim.fq); do 
	dirnamefq=$(dirname $fq);
	basenamefq=$(basename $(basename $fq .fwtrim.fq) .rvtrim.fq);
	cat $dirnamefq/${basenamefq}.fwtrim.fq  <(
		perl -wpe 'chomp;$_=reverse($_);tr/ATCGNatcgn/TAGCNtagcn/;$_.="\n";' $dirnamefq/${basenamefq}.rvtrim.fq) | \
	fastq2fasta.pl -| \
	collapse_reads_md.pl - SEQ > collapse/${basenamefq}.collapse_md.fa;
done
```

code for trimming adapters

```sh
for data in $(cat samplesheet.csv); do 
	CMD=$data";
	echo \$(gzip -dc KlaasPCRproducten_S1_L001_R1_001.fastq.gz| \
	perl ~/workspace/FastqManipulations/adapterTrimmer.pl -i - -a \$(perl -wpe 'chomp;\$_=reverse(\$_);tr/ATCGNatcgn/TAGCNtagcn/;' <(echo \$fwdPrimer)) -B -m 1 -o - |\
	perl ~/workspace/FastqManipulations/adapterTrimmer.pl -i - -a \$reversePrimer -A -m 1 -o - >trimAdapters/\${samplename}.rvtrim.fq ) &
	echo \$(gzip -dc KlaasPCRproducten_S1_L001_R1_001.fastq.gz| \
	perl ~/workspace/FastqManipulations/adapterTrimmer.pl -i - -a  \$fwdPrimer -A -m 1 -o - | \
	perl ~/workspace/FastqManipulations/adapterTrimmer.pl -i - -a \$(perl -wpe 'chomp;\$_=reverse(\$_);tr/ATCGNatcgn/TAGCNtagcn/;' <(echo \$reversePrimer )) -B -m 1 -o - > trimAdapters/\${samplename}.fwtrim.fq) &" ; 
	echo $CMD;
done
```
copypase/pipe to bash the returned commands are good for trimming

The data
========

contruct examples for the visual people. bold == adapters and in between are 33 nt barcode + linkers

<font size="5">

#8-eGFPfwd3L

CTTGTACAGCTCGTCCATGCCTTGGACGGATT

**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCTGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATG

**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGT

**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTC

**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

**CTTGTACAGCTCGTCCATGCC**TTGGATCTGTTGATCTGAAGAGCGGATCAGCAGTAATGCCGAAACCCAAATCCTATGCCGTATT

#eGFPfwd3-193

CTTGTACAGCTCGTCCATGCCAGAGATGA

**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCG

**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT

**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCG

#18-eGFPfwd3L

CTTGTACAGCTCGTCCATGCCATAAATTCGAT

**CTTGTACAGCTCGTCCATGCC**ATAAATTCGATATCTCTAGACCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATATCGTATG

**CTTGTACAGCTCGTCCATGCC**ATAAATTCGATATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

**CTTGTACAGCTCGTCCATGCC**ATAAATTCGATATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

#17-eGFPfwd3E

CTTGTACAGCTCGTCCATGCCGAGAATATTCTTG

**CTTGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTC

**CTTGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGGAATGCCGAGACCGG

**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTC



#made with

terpstramm@terpstramm-HP-EliteBook-Folio-9470m:/media/terpstramm/My_Book/largeDataByProject/PCRproductKlaasMiseq$ gzip -dc KlaasPCRproducten_S1_L001_R1_001.fastq.gz  | head -400 | grep --color=always -P "TGGGGGATCCTCACTGGCCGTC|CTGGTACAGCTCGTCCATGCC|CTTGTACAGCTCGTCCATGCC|CAGCTCGTCCATGCCTTG|CTTGTACAGCTCTTCC|TGGGGGATCCTCACCGGCCGTC|CTTGTACAGCTCGTCCATGAT|CTTGTACAGCTCGTCTATGCCGTC|CTTGTCAACCTCGTCCTTGCC|CTTGTACAGCTCGTCCATGAT|CTTGTACAGCTCGTCTATGCC" |aha --line-fix > grep.output.html

#######################################################################################################################################################

AGAGATAATA**TGGGGGATCCTCACCGGCCGTC**ATCCAATTTGAGTTAAATCGTTGACTTTGTTTGCCTTA**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGAACCGATCTCGT

CGAGATTATTA**TGGGGGATCCTCACTGGCCGTC**CTTCATCATGCGTTACTCCGGGGACTACGTGATCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATC

AGAGATAATA**TGGGGGATCCTCACTGGCCGTC**GTTCAGTTTGATGTATAACGTTAACTATGTCGGCCTTA**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

AGAGATA**TGGGGGATCCTCACTGGCCGTC**TGGCACCGTGAAATATAACGGCCACATCGTTTACCTTA**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT

AGAGAT**TGGGGGATCCTCACTGGCCGTC**CTTACGTTGAGCTACTGCGTTGACTAGGTATGTTCCTTA**CTTGTACAGCTCGTCCATGCC**ATGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT

CGAGAT**TGGGGGATCCTCACTGGCCGTC**GCACAAGTTGCATTAATGCGAGCACCAGGTCACCCTTA**CTTGTACAGCTCGTCCATGCC**ATAAATTCGATATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATG

AGAGAT**TGGGGGATCCTCACTGGCCGTC**CTCCCTCTTGCTCTAATCCGTTAACTTCGTCATCCTTA**CTTGTACAGCTCGTCCATGCC**TGAGATGAATCTTCTCTTCCTCAGAACGGTTCATCAGGGGAATGCCGAGACCGATCTCGTATGC

AGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**AGGCATAATGGTTAATACGCCAACGTCGTTATCCTTA**CTGGTACAGCTCGTCCATGCC**AAAATCGATATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGT

CGAGAT**TGGGGGATCCTCACTGGCCGTC**ATACAACATGCCATAGGCCGGCAACATTGTAATCCTTA**CTTGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGT

AGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**CCTCAATGTGGCGTAAATCGTGCACTCCGTGTCCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATC

CGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**TTGCATACTGGACTATAACGATTACATTGTAAGCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCGATATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATC

AGAGATA**TGGGGGATCCTCACTGGCCGTC**GATCAGAGTGTAGTAGTTCGGCCGTCGTAGTCCTTA**CTTGTACAGCTCGTCCATGCC**ATAAATTCGATATCTCTAGATCGGACGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGC

AGAGATA**TGGGGGATCCTCACTGGCCGTC**TACCACTTTGTCCTACAACGACAACAACGTTGCCCTTA**CTTGTACAGCTCGTCCATGCC**TTGGAATATCTCTATATCGGAAGAGCGATTCAGCAGGAATGCCGAGACCGATCTCGTATGCCG

CGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**CTTACCATGGAGTAGGTCGATTACTCTCTGCATTCCTTA**CTTGTACAGCTCTTCC**ATGCCGAGAATATTCTTGATCTCTAGATCGGACGAGCGGTTCAGCAGGAATGCCGAGACCGAT

AGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**AAGCACTTTGTGATAATTCGGGAACGTCGTGGTCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATC

CGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**CTTCATGATGCATTAACTCGATTACGTAGTGGCCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATC

AGAGATA**TGGGGGATCCTCACTGGCCGTC**ATCCAGTCTGTGTTATTGCGCTTACCTCGTGCACCTTA**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT

AGAGATA**TGGGGGATCCTCACTGGCCGTC**ATCCAGGTTGCATTACACCGTAGACCTAGTGACCCTTA**CTTGTACAGCTCGTCCATGCC**ATAAATTTCGATATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGT

AGAGAT**TGGGGGATCCTCACTGGCCGTC**GTCCAAAGTGACATAACTCGCGTACTCCGTTAACCTTA**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGACGAGCGGTTCAGCAGGAATGCCGAGACCGATATCGTATGCCG

AGAGATAATA**TGGGGGATCCTCACTGGCCGTC**CTTCATGATGCATTAACCCGATTACGTAGTGGTCCTTA**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

CGAGATAATA**TGGGGGATCCTCACTGGCCGTC**GCCCAATGTGCGGTATCTCGTTAACCATGTGGACCTTA**CTTGTACAGCTCGTCTATGCC**AGAGATGAATCTCTAGATCCGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT

CGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**ATTCATCATGCGGTCGTCGACTACATCGTTGGCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAAATGCCGAGACCGATC

AGAGATTAATA**TGGGGGATCCTCACTGGCCGTC**CATCACGCTGAAGTAGACCGCCCACGACGTACTCCTTA**CTGGTACAGCTCGTCCATGCC**GAGAATATTCTTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCT

CGAGATTATA**TGGGGGATCCTCACTGGCCGTC**CTTCGTCTGTCGTAATTCGTGGAACTTGTCGGTTCCTTC**CTTGTCAACCTCGTCCTTGCC**TTGGCCGGTTTTTCTCTAGTTCGGAAAGCGGGTTCGCAAGGAATGCCGAAACCGATTTC

CGAGATA**TGGGGGATCCTCACTGGCCGTC**CTTAGGCTGTGGTAATTCGTGGACCTTGTCGGTTCCTTA**CTTGTACAGCTCGTCCATGCC**GAGAATATTCTTGATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCG

AGAGATA**TGGGGGATCCTCACTGGCCGTC**GTGCAGTATGACCTAAAACGTGAACCACGTCAACCTTA**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAAGAATGCCGAGACCGATCTCGTAT

CGAGATA**TGGGGGATCCTCACTGGCCGTC**TAACACCTTGCAGTATGGCGGGGACTCCGTTCCCCTTACCTTTTACAGCTCTTCCTTGCCTTGGTCGGCTTGTTTCTTAGACCGGCGAACGGGTTCACCGGGAATGCCGGGCCCGTTTCCGT

AGAGATA**TGGGGGATCCTCACTGGCCGTC**CTTCACGTTATGTATGACCGATCCCCTGTCTCTCCTTTCTTTTACCACTCCTCCCTTCCCTTGACCGATCTCCTCTTCAGCAGAGCGGCTCATCAGCAATGCTGCCACCAACCTCGTCTGAC

AGAGATAATA**TGGGGGATCCTCACTGGCCGTC**GGACAACATGTGTTATTGCGAAAACCACGCACTCCTTA**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT

CGAGAT**TGGGGGATCCTCACTGGCCGTC**CTTACTCTGGTTTACCGCGAATACTTTGTGAGTTCCTTA**CTTGTACAGCTCGTCCATGCC**AGAGATGAATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCC

AGAGATA**TGGGGGATCCTCACTGGCCGTC**ATACAATATGCCATATGTCGCCCACGGTGTCAGCCTTA**CTTGTACAGCTCGTCCATGCC**TTGGACGGATTATCTCTAGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTAT
</font>
